### ISSUE TITLE
perf regression for doc generation from rollup 90416

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#90512]
Spawned off of https://github.com/rust-lang/rust/pull/90416#issuecomment-955569617

[34 doc benchmarks regressed by 1% or more.](https://perf.rust-lang.org/compare.html?start=2b643e987173b36cb0279a018579372e31a35776&end=9f13083542cb2b9fce83ed8a50238e4a6386820f)

**Comment by pnkfelix** [comment]
(spoke to @GuillaumeGomez , who hypothesizes based on the contents of Pull Request #90416 that this is a spurious regression.)

**Comment by camelid** [comment]
I wonder if it's possible that something in rustc changed that is only used by rustdoc, so the regression only showed up in doc builds?

**Comment by camelid** [comment]
I looked through the rollup diff. None of the compiler changes (and certainly not the rustdoc) changes jump out at me as potentially causing this regression. #90377 doesn't seem likely either, since I don't think it'd affect rustdoc differently from the compiler.

However, I wonder if the regression could be due to #89876. It makes some `core` internal macros more complex and adds attributes to a bunch of `core` traits, which could be slowing rustdoc down (especially since impls and traits are one of the slowest parts of rustdoc). It still seems like a bit of a stretch though to cause such huge regressions.

Also, the most significant regressions are for doc builds, but there are non-trivial regressions on check (and maybe other) builds too:

![image](https://user-images.githubusercontent.com/37223377/139949244-37f9bcea-5aa0-4601-a5a0-ecda8b60127e.png)

**Comment by jyn514** [comment]
I wonder if it makes sense not to investigate too heavily for helloworld regressions. Rustdoc has lots of global analysis which I've been meaning to look at to see if it's truly necessary, but I don't think there's much we can do to avoid regressions like this in the short-term.

**Comment by camelid** [comment]
> I wonder if it makes sense not to investigate too heavily for helloworld regressions.

I agree about helloworld, but real-world crates regressed a lot too. E.g., ripgrep regressed by almost 2%.

**Comment by camelid** [comment]
Looking at the query diff view, `build_extern_trait_impl` and `metadata_decode_entry_item_attrs` both seem to have regressed by several percentage points.
